File: lib/emerald/nodes/text_literal.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'treetop'
require_relative 'node'

# A long block of text literal, with variable templating
class TextLiteral < Node
  def to_html(context)
    body
      .elements
      .map do |element|
        if element.is_a?(Variable)
          element.to_html(context)
        else
          unescape element.text_value
        end
      end
      .join('')
      .rstrip
  end

  def unescape(text)
    text.gsub(/\\(.)/, '\1')
  end
end

File: cpp/src/nodes/text_literal.cpp
#include "text_literal.hpp"
#include <boost/algorithm/string/join.hpp>

TextLiteral::TextLiteral(NodePtrs body) : body(body) {}

std::string TextLiteral::to_html(Json &context) {
  std::vector<std::string> converted;
  std::transform(body.begin(), body.end(), converted.begin(),
      [&](NodePtr element) -> std::string { return element->to_html(context); });
  return boost::algorithm::join(converted, "");
}
